### PR TITLE
Parse info.json to set module defaults

### LINKED ISSUE #628: Add a util function which loads defaults from `info.json` files
### Please describe the problem (or idea)

In https://github.com/publiclab/image-sequencer/pull/607 the default values for some modules were changed in the `info.json` files but subsequent changes were not made to the `module.js` files which led to issues like https://github.com/publiclab/image-sequencer/pull/616 in which the html renders the default value but the sequencer doesn't 

### Possible Solution

We could add a new util function in `src/util` directory which will load all the default values from `./info.json` file and return an object based on those. This function can be used in any `module.js` file if required like for e.g 
```
const defaults = require('../../util/getDefaults.js')();
```

This will be helpful as all the defaults can be managed from the module `info.json` files and there will be no need to manage more than a single file.

### PR DESCRIPTION
**Pull request by Mridul97**
Fixes #628 

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [ ] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `rake test`
* [ ] code is in uniquely-named feature branch and has no merge conflicts
* [ ] PR is descriptively titled
* [ ] ask `@publiclab/reviewers` for help, in a comment below

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software

Thanks!

### CODE REVIEW

**Comment by Mridul97** [comment]
@jywarren Please have a look! right now, have implemented it for 2 modules only, please if this is a fine way to implement the functionality, then I will add for all the modules! :)

**Comment by jywarren** [comment]
Hi, this is really cool. I'm wondering if we could do this fewer times by including it deeper into the setup code for each module? Maybe somewhere in here? 

https://github.com/publiclab/image-sequencer/blob/2e9689b905e58a41e7604ed3d47f4b19345411a6/src/ImageSequencer.js#L358-L371

**Comment by Mridul97** [comment]
Sure, I will try it out. Thanks!

**Comment by HarshKhandeparkar** [comment]
@Mridul97 instead of hardcoding the paths in the function we could pass the paths to the function instead? Because if we hardoce them we will have to manage them too when a new dunction is added.

**Comment by Mridul97** [comment]
@HarshKhandeparkar that would give an error because we need to pass static string in require() we cannot pass a variable as require() statements are executed at compile time!

**Comment by HarshKhandeparkar** [comment]
@Mridul97 What I meant was to pass a path to the require() statement

**Comment by HarshKhandeparkar** [comment]
I think a better alternative would be to require `info.json` in `module.js` and pass that?

**Comment by Mridul97** [comment]
Ok so that means instead of creating a util function, we require info.json in module.js and use it there, we can do that I think!

**Comment by HarshKhandeparkar** [comment]
> Ok so that means instead of creating a util function, we require info.json in module.js and use it there, we can do that I think!

No, I meant we could create a util function which parses the defaults from a JSON file and takes the file as an argument.

**Comment by Mridul97** [comment]
> > Ok so that means instead of creating a util function, we require info.json in module.js and use it there, we can do that I think!
> 
> No, I meant we could create a util function which parses the defaults from a JSON file and takes the file as an argument.

That's what I was saying, we cannot do that because require() cannot take a variable as an input, it will throw a error, or maybe I am not getting you :thinking:

**Comment by HarshKhandeparkar** [comment]
like for e.g.

```js

const defaults = require('../../util/getDefaults.js')(require('./info.json');

```

**Comment by Mridul97** [comment]
> like for e.g.
> 
> ```js
> const defaults = require('../../util/getDefaults.js')(require('./info.json');
> ```

oh! okay that I will see to it! Thanks!

**Comment by Mridul97** [comment]
> Hi, this is really cool. I'm wondering if we could do this fewer times by including it deeper into the setup code for each module? Maybe somewhere in here?
> 
> [image-sequencer/src/ImageSequencer.js](https://github.com/publiclab/image-sequencer/blob/2e9689b905e58a41e7604ed3d47f4b19345411a6/src/ImageSequencer.js#L358-L371)
> 
> Lines 358 to 371 in [2e9689b](/publiclab/image-sequencer/commit/2e9689b905e58a41e7604ed3d47f4b19345411a6)
> 
>  } else if (options.func && options.info) { 
>    // passed in options object 
>    this.modules[name] = [ 
>      options.func, options.info 
>    ]; 
>   
>  } else if (options.path && !this.inBrowser) { 
>    // load from path(only in node) 
>    const module = [ 
>      require(`${options.path}/Module.js`), 
>      require(`${options.path}/info.json`) 
>    ]; 
>    this.modules[name] = module; 
>  }

@jywarren I didn't quite get how can we do it here, could you please elaborate a little more, Thanks!

**Comment by Mridul97** [comment]
@jywarren I have implemented the functionality in a better way, Please have a look!

File: src/modules/Brightness/Module.js [hunk 1]
```diff
@@ -1,15 +1,18 @@
...
 
     var output;
+    const defaults = info.inputs;
+
+
 
     function draw(input,callback,progressObj){
 
-        options.brightness = parseInt(options.brightness) || 100;
+        options.brightness = parseInt(options.brightness) || defaults.brightness.default;
```
Comment (jywarren) [inline src/modules/Brightness/Module.js]: Oh, this is nice. I have 2 suggestions. First, could we actually just more directly reference `info.inputs.brightness.default` right here?

File: src/modules/Brightness/Module.js [hunk 2]
```diff
@@ -1,15 +1,18 @@
 /*
 * Changes the Image Brightness
 */
-
+const info = require('./info.json');
```
Comment (jywarren) [inline src/modules/Brightness/Module.js]: Second idea - could we combine this with the `const defaults` line to have an absolute minimum of additions required in a module to access the defaults?

**Comment by jywarren** [comment]
For example, see how we can pass in other stuff in the `options` in the docs here:

https://github.com/publiclab/image-sequencer/blob/master/CONTRIBUTING.md#options

Just an idea!

**Comment by Mridul97** [comment]
@jywarren I will try to pre-set the defaults in the options parameter outside the module, Thanks for the suggestions!

File: src/modules/Brightness/Module.js (hunk 1 above)
Comment (Mridul97) [reply src/modules/Brightness/Module.js]: I will make the changes! Thanks!

File: src/modules/Brightness/Module.js (hunk 2 above)
Comment (Mridul97) [reply src/modules/Brightness/Module.js]: I will make the changes! Thanks!

**Comment by Mridul97** [comment]
> OK, awesome! Just a few notes and suggestions.
> 
> I'm also wondering, are we able to pre-set the defaults in the options parameter /outside/ the module? Maybe... here? Not sure, maybe we need input from @tech4GT --
> 
> [image-sequencer/src/ImageSequencer.js](https://github.com/publiclab/image-sequencer/blob/2e9689b905e58a41e7604ed3d47f4b19345411a6/src/ImageSequencer.js#L345)
> 
> Line 345 in [2e9689b](/publiclab/image-sequencer/commit/2e9689b905e58a41e7604ed3d47f4b19345411a6)
> 
>  sequencer.addSteps(stepObj.name, stepObj.options); 
> If not, we can just merge this. But perhaps this could be a broader feature improvement to consider as a followup?

@jywarren @tech4GT I am almost done but am not able to access options in addSteps() function, if we get the access to options in addSteps() function, this issue will get solved, could you please help with this? :)

**Comment by jywarren** [comment]
@publiclab/is-reviewers perhaps you could offer some support here? This will be a nice PR and it's almost there. I think we should be able to intercept options and pass them in but it needs a careful look at the code. Thanks!!!

**Comment by HarshKhandeparkar** [comment]
@jywarren instead of doing this directly, my idea was to add a util function which iterates through all inputs and provides an object of defaults. This can be helpful as defaults can be directly obtained like maybe

```js
var defaults = require('utilFunc.js')(require('./info.json'))
var input1 = defaults['input1'],
var output1 = defaults.output1
// Or even
var { in1, in2, out1, out2 } = defaults;
```

This can help if there are a lot of inputs and outputs. Thoughts? @Mridul97

**Comment by jywarren** [comment]
Hmm, so what would usage look like, @harshkhandeparkar?

**Comment by jywarren** [comment]
Trying to retrigger the build!

**Comment by jywarren** [comment]
Hmm, @publiclab/is-reviewers let's try to problem solve this -- i wasn't able to figure it out -- now https://github.com/publiclab/image-sequencer/pull/734 is depending on it, and this would really reduce redundancy in the whole lib and make setting defaults much easier!

**Comment by jywarren** [comment]
Just reformatted a bit but also to see it run Travis again.

**Comment by HarshKhandeparkar** [comment]
The travis seems to be stuck again. It was stuck for one of my PRs as well. Even Vibhor's PR. The only thing he did was that he pushed a changed travis file which also tested node 4. It worked. He reverted by removing node 4 from travis file again and it worked too! They were just do/undo commits. Next day, I pushed a commit merging with main on all of my PRs. The PR in which travis was stuck was also working. This might be a bug with travis. Maybe the tests can be run locally and force merged instead of waiting for travis if it doesn't work. Thoughts @jywarren ?

**Comment by HarshKhandeparkar** [comment]
Also the branch seems to have some conflicts. Maybe that's the problem??

**Comment by jywarren** [comment]
Yes, perhaps try resolving the conflicts, rebasing and force pushing a new commit? Potentially if that doesn't work, opening a new clean PR?

**Comment by HarshKhandeparkar** [comment]
@jywarren we still don't know what is causing this to happen. Maybe it's a travis bug. If none of the proposed workarounds work then it can be reported to travis people maybe? I have checked their repo. Other people had this bug like [this](https://github.com/travis-ci/travis-ci/issues/7602) one or [this](https://github.com/travis-ci/travis-ci/issues/10294) one, even [this](https://github.com/travis-ci/travis-ci/issues/10204) but all the issues are closed and they say that it worked later on somehow. I haven't read all the discussions but it seems like it is working. Thoughts? @jywarren

**Comment by HarshKhandeparkar** [comment]
Seems like they have a [forum thread](https://travis-ci.community/t/known-issue-travis-ci-reports-expected-waiting-for-status-to-be-reported-on-the-github-status-api-but-the-status-never-arrives/1154) discussing this. @jywarren

**Comment by HarshKhandeparkar** [comment]
@jywarren I went through the forum thread and they have provided a remedy. Some repo settings need to be changed and this probably only occurs for protected branches.

File: src/modules/Blend/Module.js [hunk 3]
```diff
@@ -1,7 +1,9 @@
 module.exports = function Dynamic(options, UI, util) {
 
-    options.func = options.func || "function(r1, g1, b1, a1, r2, g2, b2, a2) { return [ r1, g2, b2, a2 ] }";
-    options.offset = options.offset || -2;
+    var defaults = require('./../../util/getDefaults.js')(require('./info.json'));
+
+    options.func = options.func || defaults.func;
```
Comment (HarshKhandeparkar) [inline src/modules/Blend/Module.js]: Isn't this supposed to be 
`options.blend` instead of `options.func`? Also the tests are failing

Comment (Mridul97) [reply src/modules/Blend/Module.js]: Yes, just made the changes. Thanks!

**Comment by Mridul97** [comment]
@jywarren @HarshKhandeparkar I have rebased the branch and made changes! Please have a look!

**Comment by jywarren** [comment]
This is fantastic. Thanks everyone for the great problem solving!

**Pull request merged**